diff --git a/src/Plugins/Qt/QTMPipeLink.cpp b/src/Plugins/Qt/QTMPipeLink.cpp
index 9b1cc935..790e2d22 100644
--- a/src/Plugins/Qt/QTMPipeLink.cpp
+++ b/src/Plugins/Qt/QTMPipeLink.cpp
@@ -49,7 +49,14 @@ bool
 QTMPipeLink::launchCmd () {
   if (state () != QProcess::NotRunning) killProcess (1000);
   //FIXME: is UTF8 the right encoding here?
+#if (QT_VERSION >= QT_VERSION_CHECK(5,1,0))
+  QStringList tokens = QProcess::splitCommand(utf8_to_qstring(cmd));
+  QProcess::setProgram(tokens.takeFirst());
+  QProcess::setArguments(tokens);
+  QProcess::start();
+#else
   QProcess::start(utf8_to_qstring(cmd));
+#endif
   bool r= waitForStarted ();
   if (r) {
     connect (this, SIGNAL(readyReadStandardOutput ()), SLOT(readErrOut ()));
diff --git a/src/System/Link/socket_notifier.cpp b/src/System/Link/socket_notifier.cpp
index 187682ba..47dbbcee 100644
--- a/src/System/Link/socket_notifier.cpp
+++ b/src/System/Link/socket_notifier.cpp
@@ -8,7 +8,7 @@
 * It comes WITHOUT ANY WARRANTY WHATSOEVER. For details, see the file LICENSE
 * in the root directory or <http://www.gnu.org/licenses/gpl-3.0.html>.
 ******************************************************************************/
-#ifndef QTTEXMACS
+#if !(defined (QTTEXMACS) && defined (QTPIPES))
 
 #include "config.h"
 
diff --git a/src/Texmacs/Texmacs/texmacs.cpp b/src/Texmacs/Texmacs/texmacs.cpp
index 2733d8f9..3c8cc691 100644
--- a/src/Texmacs/Texmacs/texmacs.cpp
+++ b/src/Texmacs/Texmacs/texmacs.cpp
@@ -476,7 +476,7 @@ TeXmacs_main (int argc, char** argv) {
              (s == "-x") || (s == "-execute") ||
              (s == "-log-file") ||
              (s == "-build-manual") ||
-             (s == "-reference-suite") || (s == "-test-suite")) {}
+             (s == "-reference-suite") || (s == "-test-suite")) {i++;}
   }
   if (install_status == 1) {
     if (DEBUG_STD) debug_boot << "Loading welcome message...\n";
diff --git a/src/Plugins/Qt/qt_gui.cpp b/src/Plugins/Qt/qt_gui.cpp
index 1aa0500e..9765def0 100644
--- a/src/Plugins/Qt/qt_gui.cpp
+++ b/src/Plugins/Qt/qt_gui.cpp
@@ -71,9 +71,11 @@ Q_IMPORT_PLUGIN(qico)
 Q_IMPORT_PLUGIN(qsvg)
 #endif
 
-#ifdef WIN32 
-Q_IMPORT_PLUGIN(QWindowsIntegrationPlugin)
-#endif
+
+//#ifdef OS_MINGW 
+//Q_IMPORT_PLUGIN(QWindowsIntegrationPlugin)
+//#endif
+
 #ifdef QT_MAC_USE_COCOA
 Q_IMPORT_PLUGIN(QCocoaIntegrationPlugin)
 #endif
diff --git a/src/Plugins/Windows/spawn.cpp b/src/Plugins/Windows/spawn.cpp
index 36a50b61..223d5487 100644
--- a/src/Plugins/Windows/spawn.cpp
+++ b/src/Plugins/Windows/spawn.cpp
@@ -10,9 +10,10 @@
 ******************************************************************************/
 
 #include <errno.h>
+#include <windows.h>
+
 #include "array.hpp"
 #include "spawn.hpp"
-#include <windows.h>
 
 Channel::Channel (int s):sz(s) {
  origin= -1; saved= -1; fd= -1;toBeClosed= -1; str= NULL; tid= 0; 
