--- src/texmacs-mygit/misc/m4/qt.m4	2016-07-13 13:55:21.272629929 +0200
+++ src/texmacs-mygit/misc/m4/qt.m4.mod	2016-07-13 14:36:30.372197011 +0200
@@ -8,7 +8,7 @@
 
 AC_DEFUN([HACKED_AT_WITH_QT],[
 #QT has the install dir hard coded in library so we need to fix it manually for relocatable environment
-if test -z $TMBUILDENV
+if false
 then
      QT5_AVAILABLE="no"
      SAVE_LIBS="$LIBS"
@@ -37,13 +37,19 @@
      fi
 else
     # windows part
-    QT_FRAMEWORKS_PATH=/Qt
-    QT_PATH=$QT_FRAMEWORKS_PATH/bin
-    QT_PLUGINS_PATH=$QT_FRAMEWORKS_PATH/plugins
-    QT_LIBS="-L/Qt/lib -lmingw32 -lQtGui4 -lQtCore4 -lpoppler-qt4"
+    QMAKE="qmake"
+    QT_FRAMEWORKS_PATH=`$QMAKE -query QT_INSTALL_DATA`
+    #QT_FRAMEWORKS_PATH=/Qt
+    QT_PATH=`$QMAKE -query QT_INSTALL_BINS`
+    QT_PLUGINS_PATH=`$QMAKE -query QT_INSTALL_PLUGINS`
+    QT_LIB_PATH=`$QMAKE -query QT_INSTALL_LIBS`
+    QT_LIBS="-L$QT_LIB_PATH -lmingw32 -lQtGui4 -lQtCore4 -lpoppler-qt4"
+    #QT_LIBS="-L/Qt/lib -lmingw32 -lQtGui4 -lQtCore4 -lpoppler-qt4"
     MOCFLAGS="-DUNICODE -DQT_LARGEFILE_SUPPORT -DQT_DLL -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_CORE_LIB -DQT_HAVE_MMX -DQT_HAVE_3DNOW -DQT_HAVE_SSE -DQT_HAVE_MMXEXT -DQT_HAVE_SSE2 -DQT_THREAD_SUPPORT"
     QT_DEFINES="-DUNICODE -DQT_LARGEFILE_SUPPORT -DQT_DLL -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_CORE_LIB -DQT_HAVE_MMX -DQT_HAVE_3DNOW -DQT_HAVE_SSE -DQT_HAVE_MMXEXT -DQT_HAVE_SSE2 -DQT_THREAD_SUPPORT"
-    QT_CPPFLAGS="$QT_DEFINES -I/Qt/include -I/Qt/include/Qt -I/Qt/include/QtCore -I/Qt/include/QtGui"
+    QT_INCLUDE_DIR=`$QMAKE -query QT_INSTALL_HEADERS`
+    QT_CPPFLAGS="$QT_DEFINES -I$QT_INCLUDE_DIR -I$QT_INCLUDE_DIR/Qt -I$QT_INCLUDE_DIR/QtCore -I$QT_INCLUDE_DIR/QtGui"
+    #QT_CPPFLAGS="$QT_DEFINES -I/Qt/include -I/Qt/include/Qt -I/Qt/include/QtCore -I/Qt/include/QtGui"
     QT_CXXFLAGS="-O2 -frtti -fexceptions -mthreads -Wall $QT_DEFINES"
     QT_CFLAGS="-O2 -Wall $QT_DEFINES"
     MOCFLAGS=$QT_DEFINES
