Index: configure.in
===================================================================
--- configure.in	(revision 10509)
+++ configure.in	(working copy)
@@ -91,11 +91,6 @@
 AC_CPLUSPLUS_STACK
 
 #--------------------------------------------------------------------
-# Checks for Guile
-#--------------------------------------------------------------------
-LC_GUILE
-
-#--------------------------------------------------------------------
 # Checks for standard functionality
 #--------------------------------------------------------------------
 
@@ -103,6 +98,7 @@
 AC_CHECK_TYPES(FILE)
 AC_CHECK_TYPES(intptr_t)
 AC_CHECK_TYPES(time_t)
+AC_CHECK_TYPES([struct timespec])
 AC_PATH_X
 AC_PATH_XTRA
 AC_CHECK_HEADERS(pty.h util.h)
@@ -124,6 +120,11 @@
 LC_FREETYPE
 
 #--------------------------------------------------------------------
+# Checks for Guile
+#--------------------------------------------------------------------
+LC_GUILE
+
+#--------------------------------------------------------------------
 # Checks for Cairo library
 #--------------------------------------------------------------------
 LC_CAIRO
Index: misc/m4/guile.m4
===================================================================
--- misc/m4/guile.m4	(revision 10509)
+++ misc/m4/guile.m4	(working copy)
@@ -49,6 +49,7 @@
 
   LC_WITH_GUILE_tmp1="$($GUILE_CONFIG link)" && dnl
   LC_WITH_GUILE_tmp1="$LC_WITH_GUILE_tmp1 $($GUILE_CONFIG compile)" && dnl
+  LC_WITH_GUILE_tmp1="$LC_WITH_GUILE_tmp1 -I$($GUILE_CONFIG info includedir)" && dnl
   LC_WITH_GUILE_tmp1="$LC_WITH_GUILE_tmp1 -I$($GUILE_CONFIG info pkgincludedir)" && dnl
   # get th version with guile-config or guile. keep the same naming with guile version 
   # ie: guile18-config -> guile18
@@ -59,15 +60,16 @@
   GUILE_VERSION=${GUILE_VERSION%.$GUILE_VERSION_TAIL} && dnl
   GUILE_DATA_PATH=$($GUILE_CONFIG info pkgdatadir)/${GUILE_VERSION} || dnl
   AC_MSG_ERROR([cannot find guile-config; is Guile installed?])
-  LC_SCATTER_FLAGS([$LC_WITH_GUILE_tmp1], [GUILE_TMP])
+  LC_SCATTER_FLAGS([$LC_WITH_GUILE_tmp1 -lintl], [GUILE_TMP])
 
   LC_SET_EMPTY_FLAGS([GUILE_TMP],[GUILE])
+LC_DUMP_FLAGS([GUILE])
   LC_CLEAR_FLAGS([GUILE_TMP])
 
   # complete include path according the library name
-  LC_GET_ARG_VALUE(GUILE_CPPFLAGS, [-I], [LC_WITH_GUILE_tmp2])
-  LC_GET_ARG_VALUE(GUILE_LIBS, [-l], [GUILE_LIB])
-  LC_APPEND_FLAG([-I$LC_WITH_GUILE_tmp2/$GUILE_LIB], [GUILE_CPPFLAGS])
+#  LC_GET_ARG_VALUE(GUILE_CPPFLAGS, [-I], [LC_WITH_GUILE_tmp2])
+   LC_GET_ARG_VALUE(GUILE_LIBS, [-l], [GUILE_LIB])
+#  LC_APPEND_FLAG([-I$LC_WITH_GUILE_tmp2/$GUILE_LIB], [GUILE_CPPFLAGS])
 
   AC_DEFUN([GUILE_LIB_NAME], [lib$GUILE_LIB])
   unset LC_WITH_GUILE_tmp1 LC_WITH_GUILE_tmp2
@@ -144,7 +146,7 @@
         libguile18) AC_DEFINE(GUILE_HEADER_18, 1, [Guile 1.8 header]) ;;
         *) AC_MSG_WARN([Strange guile header name GUILE_LIB_NAME.h]) ;;
       esac
-      LC_LINK_IFELSE([Guile],[LM_FUNC_CHECK([gh_scm2newstr])], [
+      AC_CHECK_LIB([Guile],[gh_scm2newstr], [
         g_success=1
         LC_RUN_IFELSE([Guile DOTS], [LM_GUILE_DOTS],[
           AC_DEFINE(DOTS_OK, 1, [Defined if ...-style argument passing works])
@@ -160,7 +162,7 @@
         ],[
           AC_DEFINE(guile_str_size_t, size_t, [Guile string size type])
           AC_MSG_RESULT(checking for Guile size type... size_t)
-    ])])])])
+    ])],[],[-lintl])])])
     if [[ ! $g_success ]];then 
       AC_MSG_ERROR([It seems that guile-config does not provide the right parameters.
       Consult the config.log for error details and check your guile installation])
Index: misc/m4/x11.m4
===================================================================
--- misc/m4/x11.m4	(revision 10509)
+++ misc/m4/x11.m4	(working copy)
@@ -31,7 +31,7 @@
   then unset x_includes
   else x_libraries=${x_includes/include/lib}
   fi   AX_RESTORE_FLAGS
-  LC_SCATTER_FLAGS([-I$x_includes -L$x_libraries -lXext -lX11],[X11])
+  LC_SCATTER_FLAGS([-I$x_includes -L$x_libraries],[X11])
   LC_SUBST(X11)
   LC_COMBINE_FLAGS([X11])
 ])
Index: aclocal.m4
===================================================================
--- aclocal.m4	(revision 10521)
+++ aclocal.m4	(working copy)
@@ -22,6 +22,7 @@
 m4_include([misc/m4/freetype.m4])
 m4_include([misc/m4/dlopen.m4])
 
+
 #-------------------------------------------------------------------
 # General functions
 #-------------------------------------------------------------------
@@ -45,7 +46,7 @@
 AC_DEFUN([LC_SUBST],[ 
   AC_SUBST([$1_CXX],[$$1_CXXFLAGS])
   AC_SUBST([$1_CPP],[$$1_CPPFLAGS])
-  if [[[ $enable_dumpflags =~ (^|[[:space:]])($1|ALL)($|[[:space:]]) ]]]
+  if [[[ $enable_dumpflags =~ $(echo "(^|[[:space:]])($1|ALL)($|[[:space:]])") ]]]
   then LC_DUMP_FLAGS([$1])
   fi
 # adaptation layer remove when finished
