--- a/misc/m4/poppler.m4  (révision 10815)
+++ b/misc/m4/poppler.m4  (copie de travail)
@@ -13,7 +13,9 @@

 m4_define([PopplerLibs],[-lpoppler-qt4 -lpoppler])
 AC_DEFUN([LM_POPPLER],[
-  AC_LANG_PROGRAM([[#include "poppler/qt4/poppler-qt4.h"]],[[
+  AC_LANG_PROGRAM([[#include "poppler/qt4/poppler-qt4.h"
+  #undef main
+  ]],[[
     Poppler::Document::load("Dummy");
   ]])
 ])
--- a/misc/m4/tm_windows.m4
+++ b/misc/m4/tm_windows.m4
@@ -17,8 +17,10 @@ AC_DEFUN([TM_WINDOWS],[
       AS_HELP_STRING([--enable-console], [enable windows console]))
     if test "$enableval" = yes
     then  AC_MSG_RESULT([enabling Windows console])
+#for png needed since update to Hummus 3.9
+          LC_SCATTER_FLAGS([-lpng])
     else  AC_MSG_RESULT([disabling Windows console])
-          LC_SCATTER_FLAGS([-mwindows])
+          LC_SCATTER_FLAGS([-mwindows -lpng])
     fi
   fi
 ])
--- a/src/Plugins/Pdf/PDFWriter/DocumentContext.cpp	(révision 10845)
+++ b/src/Plugins/Pdf/PDFWriter/DocumentContext.cpp	(copie de travail)
@@ -952,6 +952,10 @@
 static const std::string scSubType = "Subtype";
 static const std::string scForm = "Form";
 static const std::string scFormType = "FormType";
+static const std::string scGroup = "Group";
+static const std::string scS = "S";
+static const std::string scTransparency = "Transparency";
+
 PDFFormXObject* DocumentContext::StartFormXObject(const PDFRectangle& inBoundingBox,ObjectIDType inFormXObjectID,const double* inMatrix,const bool inUseTransparencyGroup)
 {
 	PDFFormXObject* aFormXObject = NULL;
@@ -986,8 +990,11 @@
 			mObjectsContext->EndArray(eTokenSeparatorEndLine);
 		}
         if (inUseTransparencyGroup) {
-          xobjectContext->WriteKey("Group");
-          xobjectContext->WriteHexStringValue("</S /Transparency>");
+          xobjectContext->WriteKey(scGroup);
+	  DictionaryContext* groupContext = mObjectsContext->StartDictionary();
+	  groupContext->WriteKey(scS);
+	  groupContext->WriteNameValue(scTransparency);
+	  mObjectsContext->EndDictionary(groupContext);
         }
 
 		// Resource dict 
@@ -1895,7 +1902,6 @@
 static const std::string scA = "A";
 static const std::string scBS = "BS";
 static const std::string scAction = "Action";
-static const std::string scS = "S";
 static const std::string scURI = "URI";
 EStatusCodeAndObjectIDType DocumentContext::WriteAnnotationAndLinkForURL(const std::string& inURL,const PDFRectangle& inLinkClickArea)
 {
@@ -2928,7 +2934,7 @@
 	ObjectIDType inObjectID,
 	const PDFParsingOptions& inParsingOptions)
 {
-    EStatusCode status;
+    EStatusCode status = eFailure;
     EHummusImageType imageType = GetImageType(inImagePath,inImageIndex);
         
     switch(imageType)
