--- a/misc/m4/poppler.m4  (révision 10815)
+++ b/misc/m4/poppler.m4  (copie de travail)
@@ -13,7 +13,9 @@

 m4_define([PopplerLibs],[-lpoppler-qt4 -lpoppler])
 AC_DEFUN([LM_POPPLER],[
-  AC_LANG_PROGRAM([[#include "poppler/qt4/poppler-qt4.h"]],[[
+  AC_LANG_PROGRAM([[#include "poppler/qt4/poppler-qt4.h"
+  #undef main
+  ]],[[
     Poppler::Document::load("Dummy");
   ]])
 ])
--- a/misc/m4/tm_windows.m4
+++ b/misc/m4/tm_windows.m4
@@ -17,8 +17,10 @@ AC_DEFUN([TM_WINDOWS],[
       AS_HELP_STRING([--enable-console], [enable windows console]))
     if test "$enableval" = yes
     then  AC_MSG_RESULT([enabling Windows console])
+#for png needed since update to Hummus 3.9
+          LC_SCATTER_FLAGS([-lpng])
     else  AC_MSG_RESULT([disabling Windows console])
-          LC_SCATTER_FLAGS([-mwindows])
+          LC_SCATTER_FLAGS([-mwindows -lpng])
     fi
   fi
 ])
