diff --git a/configure.in b/configure.in
index 7f0a87f..c29478c 100644
--- a/configure.in
+++ b/configure.in
@@ -22,7 +22,12 @@ AC_CANONICAL_HOST
 AC_MSG_CHECKING(current Subversion revision number)
 SVNREV=`svnversion -n . 2>/dev/null`
 SVNREV=${SVNREV/:/_}
-if { test "$SVNREV" = "" || test "$SVNREV" = "exported" ; } ; then 
+if case $SVNREV in *[[0-9]]*) false;; *) true;; esac;  #contains no digit -> invalid
+then # maybe it was a git ?
+  SVNREV=`git svn find-rev $(git rev-parse HEAD 2>/dev/null) 2>/dev/null`
+fi
+if case $SVNREV in *[[0-9]]*) false;; *) true;; esac;
+then 
   SVNREV=`cat $srcdir/SVNREV`
   AC_MSG_RESULT($SVNREV, read from $srcdir/SVNREV)
 else 
@@ -39,7 +44,7 @@ AC_SUBST(SVNREV)
 PACKAGE="TeXmacs"
 DEBIAN_NAME="texmacs"
 DEVEL_VERSION="1.99.5"
-DEVEL_RELEASE="1"         # I think we should use SVNREV here
+DEVEL_RELEASE=$SVNREV         # actually using SVNREV here
 STABLE_VERSION="1.99.5"
 STABLE_RELEASE="1"
 
diff --git a/misc/m4/sparkle.m4 b/misc/m4/sparkle.m4
index 14b4829..8d0e801 100644
--- a/misc/m4/sparkle.m4
+++ b/misc/m4/sparkle.m4
@@ -4,6 +4,7 @@ AC_DEFUN([LC_SPARKLE],[
   [with (Win)Sparkle autoupdater [ARG=no]]))
 
   SAVE_OBJCFLAGS="$OBJCFLAGS"
+  SAVE_CFLAGS="$CFLAGS"
   SAVE_LDFLAGS="$LDFLAGS"
   SAVE_LIBS="$LIBS"
   if test "$with_sparkle" = "no" -o "$with_sparkle" = "" ; then
@@ -19,13 +20,12 @@ AC_DEFUN([LC_SPARKLE],[
       AC_TRY_LINK(
   [
   #include <winsparkle.h>
-  #include <winsparkle.h>
   ],[
   win_sparkle_init();
   ],[
       AC_MSG_RESULT(yes)
       AC_DEFINE(USE_SPARKLE, 1, [Use WinSparkle library])
-      SPARKLE_CFLAGS="$CFLAGS"
+      SPARKLE_CXXFLAGS="$CFLAGS"
       SPARKLE_LDFLAGS="$LDFLAGS $LIBS"
       WINSPARKLE_PATH="$with_sparkle"
       WINSPARKLE_DLL="WinSparkle*.dll"
@@ -59,9 +59,10 @@ AC_DEFUN([LC_SPARKLE],[
     esac
   fi
 
-  OBJFLAGS="$SAVE_CPPFLAGS"
-  LDFLAGS="$SAVE_LDFLAGS"
-  LIBS="$SAVE_LIBS"
+  OBJFLAGS="$SAVE_OBJCFLAGS"
+  CFLAGS="$SAVE_CFLAGS"
+  LDFLAGS="$LDFLAGS $SAVE_LDFLAGS"
+  LIBS="$LIBS $SAVE_LIBS"
 
   LC_SUBST(SPARKLE)
   AC_SUBST(SPARKLE_FRAMEWORK_PATH)
