diff --git a/TeXmacs/doc/about/about-summary.de.tm b/TeXmacs/doc/about/about-summary.de.tm
index 5a623ae..3efbc7f 100644
--- a/TeXmacs/doc/about/about-summary.de.tm
+++ b/TeXmacs/doc/about/about-summary.de.tm
@@ -7,7 +7,7 @@
 
   <big-table|<descriptive-table|<tformat|<cwith|1|1|1|1|cell-row-span|1>|<cwith|1|1|1|1|cell-col-span|2>|<cwith|1|1|1|1|cell-halign|c>|<table|<row|<cell|GNU
   <TeXmacs>>|<cell|>>|<row|<cell|Installierte
-  Version>|<cell|<TeXmacs-version>>>|<row|<cell|Unterstützte
+  Version>|<cell|<TeXmacs-version-release|devel-release>>>|<row|<cell|Unterstützte
   Systeme>|<cell|Most GNU/<name|Linux> systems>>|<row|<cell|Copyright>|<cell|<copyright>
   1998--2002 by Joris van der Hoeven >>|<row|<cell|Lizenz>|<cell|<hlink|GNU
   General Public License|$TEXMACS_PATH/LICENSE>>>|<row|<cell|Webseiten>|<cell|<tabular|<tformat|<twith|table-valign|T>|<cwith|1|-1|1|-1|cell-lsep|0spc>|<cwith|1|-1|1|-1|cell-rsep|0spc>|<table|<row|<cell|<verbatim|http://www.texmacs.org>>>|<row|<cell|<verbatim|http://www.gnu.org/software/texmacs>>>>>>>>|<row|<cell|Kontakt>|<cell|<verbatim|contact@texmacs.org>>>|<row|<cell|Postansschrift>|<cell|<tabular|<tformat|<twith|table-valign|T>|<cwith|1|-1|1|-1|cell-lsep|0spc>|<cwith|1|-1|1|-1|cell-rsep|0spc>|<table|<row|<cell|<abbr|Dr.>
diff --git a/TeXmacs/doc/about/about-summary.en.tm b/TeXmacs/doc/about/about-summary.en.tm
index 09240c1..6e2f2a9 100644
--- a/TeXmacs/doc/about/about-summary.en.tm
+++ b/TeXmacs/doc/about/about-summary.en.tm
@@ -7,7 +7,7 @@
 
   <big-table|<descriptive-table|<tformat|<cwith|1|1|1|1|cell-row-span|1>|<cwith|1|1|1|1|cell-col-span|2>|<cwith|1|1|1|1|cell-halign|c>|<table|<row|<cell|GNU
   <TeXmacs>>|<cell|>>|<row|<cell|Installed
-  version>|<cell|<TeXmacs-version>>>|<row|<cell|Supported systems>|<cell|Most
+  version>|<cell|<TeXmacs-version-release|devel-release>>>|<row|<cell|Supported systems>|<cell|Most
   GNU/<name|Linux> systems>>|<row|<cell|Copyright>|<cell|<copyright>
   1998--2002 by Joris van der Hoeven >>|<row|<cell|License>|<cell|<hlink|GNU
   General Public License|$TEXMACS_PATH/LICENSE>>>|<row|<cell|Web
diff --git a/TeXmacs/doc/about/about-summary.es.tm b/TeXmacs/doc/about/about-summary.es.tm
index 4feed8b..525982c 100644
--- a/TeXmacs/doc/about/about-summary.es.tm
+++ b/TeXmacs/doc/about/about-summary.es.tm
@@ -6,7 +6,7 @@
   <with|language|spanish|><tmdoc-title|Sumario>
 
   <big-table|<descriptive-table|<tformat|<cwith|1|1|1|1|cell-row-span|1>|<cwith|1|1|1|1|cell-col-span|2>|<cwith|1|1|1|1|cell-halign|c>|<table|<row|<cell|GNU
-  <TeXmacs>>|<cell|>>|<row|<cell|Versión>|<cell|<TeXmacs-version>>>|<row|<cell|Sistemas
+  <TeXmacs>>|<cell|>>|<row|<cell|Versión>|<cell|<TeXmacs-version-release|devel-release>>>|<row|<cell|Sistemas
   soportados>|<cell|La mayoría de los sistemas
   GNU/<name|Linux>>>|<row|<cell|Copyright>|<cell|<copyright> 1998--2002 por
   Joris van der Hoeven >>|<row|<cell|Licencia>|<cell|<hlink|Licencia Pública
diff --git a/TeXmacs/doc/about/about-summary.fr.tm b/TeXmacs/doc/about/about-summary.fr.tm
index 7bef7be..ffcfba7 100644
--- a/TeXmacs/doc/about/about-summary.fr.tm
+++ b/TeXmacs/doc/about/about-summary.fr.tm
@@ -6,7 +6,7 @@
   <tmdoc-title|Sommaire>
 
   <big-table|<descriptive-table|<tformat|<cwith|1|1|1|1|cell-row-span|1>|<cwith|1|1|1|1|cell-col-span|2>|<cwith|1|1|1|1|cell-halign|c>|<table|<row|<cell|GNU
-  <TeXmacs>>|<cell|>>|<row|<cell|Version installée>|<cell|<TeXmacs-version>>>|<row|<cell|Systèmes
+  <TeXmacs>>|<cell|>>|<row|<cell|Version installée>|<cell|<TeXmacs-version-release|devel-release>>>|<row|<cell|Systèmes
   reconnus>|<cell|la plupart des systèmes
   GNU/<name|Linux>>>|<row|<cell|Droits d'auteur>|<cell|<copyright> 1998--2003
   de Joris van der Hoeven >>|<row|<cell|Licence>|<cell|<hlink|GNU General
diff --git a/TeXmacs/doc/about/about-summary.it.tm b/TeXmacs/doc/about/about-summary.it.tm
index 3be3d46..ae1e802 100644
--- a/TeXmacs/doc/about/about-summary.it.tm
+++ b/TeXmacs/doc/about/about-summary.it.tm
@@ -6,7 +6,7 @@
   <tmdoc-title|Sommario>
 
   <big-table|<descriptive-table|<tformat|<cwith|1|1|1|1|cell-row-span|1>|<cwith|1|1|1|1|cell-col-span|2>|<cwith|1|1|1|1|cell-halign|c>|<table|<row|<cell|GNU
-  <TeXmacs>>|<cell|>>|<row|<cell|<localize|Version>>|<cell|<TeXmacs-version>>>|<row|<cell|Sistemi
+  <TeXmacs>>|<cell|>>|<row|<cell|<localize|Version>>|<cell|<TeXmacs-version-release|devel-release>>>|<row|<cell|Sistemi
   supportati>|<cell|Principalmente sistemi
   GNU/<name|Linux>>>|<row|<cell|Copyright>|<cell|<copyright> 1998--2002 by
   Joris van der Hoeven >>|<row|<cell|Licenza>|<cell|<hlink|GNU General Public
diff --git a/TeXmacs/doc/about/about-summary.zh.tm b/TeXmacs/doc/about/about-summary.zh.tm
index 65e0107..f5d8e0c 100644
--- a/TeXmacs/doc/about/about-summary.zh.tm
+++ b/TeXmacs/doc/about/about-summary.zh.tm
@@ -6,7 +6,7 @@
   <tmdoc-title|\<#6982\>\<#8981\>>
 
   <big-table|<descriptive-table|<tformat|<cwith|1|1|1|1|cell-row-span|1>|<cwith|1|1|1|1|cell-col-span|2>|<cwith|1|1|1|1|cell-halign|c>|<table|<row|<cell|GNU
-  <TeXmacs>>|<cell|>>|<row|<cell|\<#5F53\>\<#524D\>\<#7248\>\<#672C\>\<#53F7\>>|<cell|<TeXmacs-version>>>|<row|<cell|\<#652F\>\<#6301\>\<#7684\>\<#7CFB\>\<#7EDF\>>|<cell|\<#5927\>\<#90E8\>\<#5206\>GNU/<name|Linux>\<#7CFB\>\<#7EDF\>>>|<row|<cell|\<#7248\>\<#6743\>>|<cell|<copyright>
+  <TeXmacs>>|<cell|>>|<row|<cell|\<#5F53\>\<#524D\>\<#7248\>\<#672C\>\<#53F7\>>|<cell|<TeXmacs-version-release|devel-release>>>|<row|<cell|\<#652F\>\<#6301\>\<#7684\>\<#7CFB\>\<#7EDF\>>|<cell|\<#5927\>\<#90E8\>\<#5206\>GNU/<name|Linux>\<#7CFB\>\<#7EDF\>>>|<row|<cell|\<#7248\>\<#6743\>>|<cell|<copyright>
   1998--2002 by Joris van der Hoeven >>|<row|<cell|\<#8BB8\>\<#53EF\>\<#8BC1\>>|<cell|<hlink|GNU
   General Public License|$TEXMACS_PATH/LICENSE>>>|<row|<cell|\<#9879\>\<#76EE\>\<#4E3B\>\<#9875\>>|<cell|<tabular|<tformat|<twith|table-valign|T>|<cwith|1|-1|1|-1|cell-lsep|0spc>|<cwith|1|-1|1|-1|cell-rsep|0spc>|<table|<row|<cell|<verbatim|http://www.texmacs.org>>>|<row|<cell|<verbatim|http://www.gnu.org/software/texmacs>>>>>>>>|<row|<cell|\<#8054\>\<#7CFB\>\<#6211\>\<#4EEC\>>|<cell|<verbatim|contact@texmacs.org>>>|<row|<cell|\<#8054\>\<#7CFB\>\<#5730\>\<#5740\>>|<cell|<tabular|<tformat|<twith|table-valign|T>|<cwith|1|-1|1|-1|cell-lsep|0spc>|<cwith|1|-1|1|-1|cell-rsep|0spc>|<table|<row|<cell|<abbr|Dr.>
   Joris van der Hoeven>>|<row|<cell|<abbr|Dépt.> de Mathématiques
diff --git a/TeXmacs/progs/utils/misc/updater.scm b/TeXmacs/progs/utils/misc/updater.scm
index 99a8f3f..ced24fb 100644
--- a/TeXmacs/progs/utils/misc/updater.scm
+++ b/TeXmacs/progs/utils/misc/updater.scm
@@ -52,7 +52,7 @@
         (else (noop))))
 
 (define-preferences
-  ("updater:appcast" "http://www.texmacs.org/Download/ftp/tmftp/appcast.xml"
+  ("updater:appcast" "https://raw.githubusercontent.com/slowphil/mingw-w64-texmacs/master/appcast.xml"
                      updater-notify-pref)
   ; dsa key ignored by Sparkle (set in Info.plist), unused by WinSparkle
   ("updater:public-dsa-key" "texmacs_updates_dsa_pub.pem" updater-notify-pref)
diff --git a/configure.in b/configure.in
index 5282fa5..ef87fab 100644
--- a/configure.in
+++ b/configure.in
@@ -24,7 +24,7 @@ TM_SUBVERSION
 PACKAGE="TeXmacs"
 DEBIAN_NAME="texmacs"
 DEVEL_VERSION="1.99.8"
-DEVEL_RELEASE="1"         # I think we should use SVNREV here
+DEVEL_RELEASE=$SVNREV         # actually using SVNREV here
 STABLE_VERSION="1.99.8"
 STABLE_RELEASE="1"
 
diff --git a/misc/m4/tm_subversion.m4 b/misc/m4/tm_subversion.m4
index e9a94db..e5a3ef0 100644
--- a/misc/m4/tm_subversion.m4
+++ b/misc/m4/tm_subversion.m4
@@ -11,11 +11,20 @@
 #
 #--------------------------------------------------------------------
 
+#--------------------------------------------------------------------
+# Retrieve SVN revision number from working dir or fallback file
+#--------------------------------------------------------------------
+
 AC_DEFUN([TM_SUBVERSION],[
   AC_MSG_CHECKING(current Subversion revision number)
   SVNREV=`svnversion -n . 2>/dev/null`
   SVNREV=${SVNREV/:/_}
-  if { test "$SVNREV" = "" || test "$SVNREV" = "exported" ; } ; then 
+  if test "$(echo $SVNREV | tr -cd [[:digit:]])" = "" ; 
+  then #contains no digit -> invalid
+  # maybe it was a git mirror repo? (needs git-svn)
+    SVNREV=`git svn log --oneline -1 2>/dev/null | cut -d '|' -f1`
+  fi
+  if test "$(echo $SVNREV | tr -cd [[:digit:]])" = "" ; then
     SVNREV=`cat $srcdir/SVNREV`
     AC_MSG_RESULT($SVNREV, read from $srcdir/SVNREV)
   else 
diff --git a/packages/windows/TeXmacs.iss.in b/packages/windows/TeXmacs.iss.in
index c4d9cc7..5705f24 100644
--- a/packages/windows/TeXmacs.iss.in
+++ b/packages/windows/TeXmacs.iss.in
@@ -17,7 +17,7 @@
 
 UninstallDisplayIcon={app}\TeXmacs.ico
 OutputDir=..\..\..\distr\windows
-OutputBaseFilename=@tm_devel@.@SVNREV@-win-x86
+OutputBaseFilename=TeXmacs-@DEVEL_VERSION@.@SVNREV@-win-x86
 ; SourceDir=../..
 ChangesAssociations=yes
 
@@ -34,7 +34,7 @@
 Root: HKCR; Subkey: "tmfile\shell\open\command"; ValueType: string; ValueName: ""; ValueData: """{app}\bin\texmacs.exe"" ""%1""" 
 
 [Files]
-Source: ..\..\..\distr\TeXmacs-Windows\*; DestDir: {app}; Flags: recursesubdirs createallsubdirs
+Source: ..\..\..\distr\TeXmacs-Windows\*; DestDir: {app}; Flags: ignoreversion recursesubdirs createallsubdirs
 Source: TeXmacs.ico; DestDir: {app}
 
 [Icons]
diff --git a/packages/windows/resource.rc.in b/packages/windows/resource.rc.in
index bcee342..2cf508d 100644
--- a/packages/windows/resource.rc.in
+++ b/packages/windows/resource.rc.in
@@ -31,7 +31,7 @@ BEGIN
       VALUE "OriginalFilename", "texmacs.exe\0"
       VALUE "LegalCopyright", "Licensed under GPLv3\0"
       VALUE "ProductName", "TeXmacs\0"
-      VALUE "ProductVersion", "@DEVEL_VERSION@\0"
+      VALUE "ProductVersion", "@DEVEL_VERSION@.@SVNREV@\0"
     END
   END
 
